### PR TITLE
net/coap: allow user to configure message retry macros

### PR DESCRIPTION
**Pull request by kb2ma**
### Contribution description
Allows the user to externally specify a value for several macros used for CoAP confirmable messaging. Also adds a documentation group for these user configurable macros, per #10566. 

Most of these macros are defined in RFC 7252, but I also added documentation here to make their purpose clear.

### Testing procedure
Compile the documentation. You should see:

- a new configuration group titled, "CoAP compile configurations"
- a new module within "CoAP defines", also titled "CoAP compile configurations"

### Issues/PRs references
Partially implements #10566

### CODE REVIEW

**Comment by miri64** [comment]
Aren't those RFC-provided?

**Comment by kb2ma** [comment]
COAP_ACK_TIMEOUT, COAP_RANDOM_FACTOR, and COAP_MAX_RETRANSMIT are defined in the 7252. These values all are defined as _defaults_ from the RFC. However, the RFC also says they may be customized for a deployment within certain bounds.

@miri64, what is your concern here?

**Comment by miri64** [comment]
Ok, I wasn't sure about that. Maybe amend a link to that reference (in a separate PR/commit) to the documentation. :-).

File: sys/include/net/coap.h [hunk 1]
```diff
@@ -157,20 +157,51 @@ extern "C" {
...
+#endif
+
+/** @brief   Maximum number of retransmissions for a confirmable request */
+#ifndef COAP_MAX_RETRANSMIT
 #define COAP_MAX_RETRANSMIT     (4)
+#endif
+/** @} */
+
+/**
+ * @name    Timing parameters
```
Comment (miri64) [inline sys/include/net/coap.h]: Aren't the once above not also timing parameters still?

**Comment by miri64** [comment]
> @miri64, what is your concern here?

Since the RFC actively states they are supposed to be configurable, none.

File: sys/include/net/coap.h (hunk 1 above)
Comment (kb2ma) [reply sys/include/net/coap.h]: Yes, they are. I can't find a way to use the same name "Timing parameters" to keep the list of timing parameters that also are compile configurations _separate_ from those timing parameters  that are not also compile configurations.

If I use a different name for the timing parameters that also are compile configurations, like "Timing parameters compile configurations", then the lists stay separate. I'd appreciate any suggestions on the best approach.

Comment (miri64) [reply sys/include/net/coap.h]: They could stay separate, but if the others are also timing parameters, maybe "Fixed timing parameters"?

Comment (kb2ma) [reply sys/include/net/coap.h]: Thanks for the suggestion. I created a separate commit for that.

**Comment by kb2ma** [comment]
Created a separate commit for the RFC reference for configurability. Also added the doc label for the supergroup. Ready for hopefully final review.

File: sys/include/net/coap.h [hunk 2]
```diff
@@ -156,21 +156,62 @@ extern "C" {
...
+/**
+ * @defgroup net_coap_conf    CoAP compile configurations
+ * @ingroup  net_coap
+ * @ingroup  config
+ * @{
+ */
 /**
  * @name    Timing parameters
+ *
+ * These parameters are defined as configurable in RFC 7252, section 4.8.1.
```
Comment (miri64) [inline sys/include/net/coap.h]:
```suggestion
  * These parameters are defined as configurable in [RFC 7252, section 4.8.1]
  * (https://tools.ietf.org/html/rfc7252#section-4.8.1).
```
?

File: sys/include/net/coap.h [hunk 3]
```diff
@@ -156,21 +156,62 @@ extern "C" {
...
  * @{
  */
+/**
+ * @brief    Timeout in seconds for a response to a confirmable request
+ *
+ * This value is for the response to the *initial* confirmable message. The
+ * timeout doubles for subsequent retries. To avoid synchronization of resends
+ * across hosts, the actual timeout is chosen randomly between COAP_ACK_TIMEOUT
+ * and (COAP_ACK_TIMEOUT * COAP_RANDOM_FACTOR).
```
Comment (miri64) [inline sys/include/net/coap.h]: Add `@ref` before the mentioning of these so they are linked.

File: sys/include/net/coap.h (hunk 2 above)
Comment (kb2ma) [reply sys/include/net/coap.h]: Fixed in a9e49fb.

File: sys/include/net/coap.h (hunk 3 above)
Comment (kb2ma) [reply sys/include/net/coap.h]: Fixed in 75cc1f7, and fixed a couple of other macro references as well.

**Comment by kb2ma** [comment]
Squashed, looks good to me.

**Pull request merged**